lib/excon/socket.rb

```diff
@@ -172,20 +187,21 @@ def connect
       end
     end
 
+    def add_to_read_buffer(str)
+      @read_buffer << str
+      @eof = false
```

Comment:
Since `@eof` is initialized as false, I would expect it to still be false here. Is there a case where it would have been set to something other than false where it would then need changed back here? Otherwise, it may be better to not do this (to avoid complexity and/or confusion in the future if we don't really need to do this).

Reply:
Yeah, we're trying to preserve the invariant: @eof is true only if there is no buffered data which we could return.

For example, imagine this series of events, assuming chunk_size = 1 MB:

1. Socket is initialized, @eof = false
2. readline is called
3. readline calls read_nonblock, which reads 800K and gets EOF. Now @eof is true, and read_nonblock returns an 800K block of data
4. readline finds the first newline in the block, and it's only 100 bytes in
5. readline calls add_to_read_buffer to put the remaining 799.9K back in the buffer
6. Now `@eof` has to be false, since a call to read_nonblock could return successfully

Reply:
Got it. This relates to the new treatment of the buffer as well, I think it's all coming together in my head.